devdoc/interlocked_requirements.md

```diff
@@ -0,0 +1,298 @@
+MOCKABLE_FUNCTION(, void*, interlocked_exchange_pointer, volatile void*, target, void*, value);
+MOCKABLE_FUNCTION(, int32_t, interlocked_increment, volatile int32_t*, addend);
+MOCKABLE_FUNCTION(, int16_t, interlocked_increment_16, volatile in       t16_t*, addend);
+MOCKABLE_FUNCTION(, int64_t, interlocked_increment_64, volatile int64_t*, addend);
+MOCKABLE_FUNCTION(, int32_t, interlocked_or, volatile int32_t*, destination, int32_t, value);
+MOCKABLE_FUNCTION(, int16_t, interlocked_or_16, volatile int16_t*, destination, int16_t, value);
+MOCKABLE_FUNCTION(, int64_t, interlocked_or_64, volatile int64_t*, destination, int64_t, value);
+MOCKABLE_FUNCTION(, int8_t, interlocked_or_8, volatile int8_t*, destination, int8_t, value);
+MOCKABLE_FUNCTION(, int32_t, interlocked_xor, volatile int32_t*, destination, int32_t, value);
+MOCKABLE_FUNCTION(, int16_t, interlocked_xor_16, volatile int16_t*, destination, int16_t, value);
+MOCKABLE_FUNCTION(, int64_t, interlocked_xor_64, volatile int64_t*, destination, int64_t, value);
+MOCKABLE_FUNCTION(, int8_t, interlocked_xor_8, volatile int8_t*, destination, int8_t, value);
+```
+
+## interlocked_add
+
+```c
+MOCKABLE_FUNCTION(, int32_t, interlocked_add, volatile int32_t*, addend, int32_t, value);
+```
+
+**SRS_INTERLOCKED_43_001 [** `interlocked_add` shall perform an atomic addition operation on the specified 32-bit integer values.**]**
+
+
+
+## interlocked_and
+
+```c
+MOCKABLE_FUNCTION(, int32_t, interlocked_and, volatile int32_t*, destination, int32_t, value);
+```
+
+**SRS_INTERLOCKED_43_002 [** `interlocked_and` shall perform an atomic AND operation on the specified 32-bit integer values.**]**
+
+## interlocked_and_16
+
+```c
+MOCKABLE_FUNCTION(, int16_t, interlocked_and_16, volatile int16_t*, destination, int16_t, value);
+```
+
+**SRS_INTERLOCKED_43_003 [** `interlocked_and_16` shall perform an atomic AND operation on the specified 16-bit integer values.**]**
+
+
+## interlocked_and_64
+
+```c
+MOCKABLE_FUNCTION(, int64_t, interlocked_and_64, volatile int64_t*, destination, int64_t, value);
+```
+
+**SRS_INTERLOCKED_43_004 [** `interlocked_and_64` shall perform an atomic AND operation on the specified 64-bit integer values.**]**
+
+## interlocked_and_8
+
+```c
+MOCKABLE_FUNCTION(, int64_t, interlocked_and_8, volatile int8_t*, destination, int8_t, value);
+```
+
+**SRS_INTERLOCKED_43_005 [** `interlocked_and_8` shall perform an atomic AND operation on the specified 8-bit integer values.**]**
+
+## interlocked_compare_exchange
+
+```c
+MOCKABLE_FUNCTION(, int32_t, interlocked_compare_exchange, volatile int32_t*, destination, int32_t, exchange, int32_t, comperand);
+```
+
+**SRS_INTERLOCKED_43_006 [** `interlocked_compare_exchange` shall perform an atomic compare-and-exchange operation on the specified values. The function compares two specified 32-bit values and exchanges with another 32-bit value based on the outcome of the comparison.**]**
```

Comment:
>`interlocked_compare_exchange` shall perform an atomic compare-and-exchange operation on the specified values. The function compares two specified 32-bit values and exchanges with another 32-bit value based on the outcome of the comparison. [](start = 29, length = 240)

I think you can break down the compare and exchange to a couple requirements (and there should be a couple of test cases anyway):

`interlocked_compare_exchange` shall atomically compare the 32-bit `destination` with the 32-bit `comparand`.

If the `destination` is the same as `comparand` then `interlocked_compare_exchange` shall store `exchange` in `destination` as part of the atomic operation.

`interlocked_compare_exchange` shall return the value of `destination` at the time of the compare.


The same goes for the other compare-and-exchange. It's nice to break the requirements down like this generally so it's clear what test cases need to be written